sktime/util_torch.py

```diff
@@ -0,0 +1,207 @@
+
+        :getter: The best value.
+        :type: float or None
+        """
+        return self._best_metric_value
+
+    @property
+    def best_metric_mode(self) -> str:
+        r""" The mode under which the metric is taken into account. Can either be 'max', i.e., larger metric values are
+        better, or 'min' for the opposite behavior.
+
+        :getter: Yields the mode.
+        :setter: Sets the mode.
+        :type: str, one of ('max', 'min')
+        """
+        return self._best_metric_mode
+
+    @best_metric_mode.setter
+    def best_metric_mode(self, value: str):
+        assert value in ('min', 'max'), f"Unknown mode {value}, supported are min and max."
+        self._best_metric_mode = value
+
+    def step(self, step: int, metric_value: float, models: Dict[str, torch.nn.Module]):
+        r""" Records a step in the checkpoint manager. This automatically prunes old checkpoints according to
+        :meth:`max_n_checkpoints`.
+
+        Parameters
+        ----------
+        step : int
+            The training step. Should be monotonically increasing during training.
+        metric_value : float
+            Value of the metric.
+        models : dict from model name to torch module
+            Dictionary of modules which should be stored in the checkpoint.
+        """
+        self._make_checkpoint(step, models, self._output_dir / f"checkpoint_{step}.ckpt")
+        self.prune_checkpoints()
+        if metric_value is not None:
+            op = min if self.best_metric_mode == 'min' else max
+            if self.best_metric_value is None or (metric_value == op(self.best_metric_value, metric_value)):
+                self._best_metric_value = metric_value
+                self._make_checkpoint(step, models, self._output_dir / f"best.ckpt")
+
+    @staticmethod
+    def _make_checkpoint(step, models: Dict[str, torch.nn.Module], outfile: Path):
+        r""" Makes the actual checkpoint. """
+        save_dict = {k: v.state_dict() for k, v in models}
+        save_dict['step'] = step
+        torch.save(save_dict, outfile)
+        return outfile
+
+    def prune_checkpoints(self):
+        r""" Prunes old checkpoints. """
+        checkpoints = list(self._output_dir.glob("*.ckpt"))
+        steps = []
+        for ckpt in checkpoints:
+            fname = str(Path(ckpt).name)
+            filtered = "".join(list(c for c in filter(str.isdigit, fname)))
+            if len(filtered) > 0:
+                steps.append(int(filtered))
+        while len(steps) > self.keep_n_checkpoints:
+            oldest = min(steps)
+            oldest_path = self._output_dir.joinpath(
+                "checkpoint_{}.ckpt".format(oldest)
+            )
+            oldest_path.unlink()
+            steps.remove(oldest)
+
+
+class Stats(object):
+    r""" Object that collects training statistics in a certain group. """
+
+    def __init__(self, group: str, items: List[str]):
```

Comment:
i've read through this a few times and i'm a little confused/maybe more docs would be good. can the user provide their own custom stats that should be calculated/tracked? what is `items`?

Reply:
Ill add an example to the docs, that should clear things up!